dlib/invoke.h

```diff
@@ -11,32 +11,153 @@ namespace dlib
+        }
+
+        template<
+            typename Base,
+            typename T,
+            typename Derived
+        >
+        constexpr auto INVOKE(
+            T Base::*pmd, //pointer to member data
+            Derived&& ref
+        )
+        noexcept(noexcept(std::forward<Derived>(ref).*pmd))
+        -> typename std::enable_if<
+                std::is_object<T>::value &&
+                std::is_base_of<Base, typename std::decay<Derived>::type>::value,
+                decltype(std::forward<Derived>(ref).*pmd)>::type
+        {
+            return std::forward<Derived>(ref).*pmd;
+        }
+
+        template<
+            typename Base,
+            typename T,
+            typename RefWrap
+        >
+        constexpr auto INVOKE(
+            T Base::*pmd, //pointer to member data
+            RefWrap&& ref
+        )
+        noexcept(noexcept(ref.get().*pmd))
+        -> typename std::enable_if<
+                std::is_object<T>::value &&
+                is_reference_wrapper<typename std::decay<RefWrap>::type>::value,
+                decltype(ref.get().*pmd)>::type
+        {
+            return ref.get().*pmd;
+        }
+
+        template<
+            typename Base,
+            typename T,
+            typename Ptr
+        >
+        constexpr auto INVOKE(
+            T Base::*pmd, //pointer to member data
+            Ptr&& ptr
+        )
+        noexcept(noexcept((*std::forward<Ptr>(ptr)).*pmd))
+        -> typename std::enable_if<
+                std::is_object<T>::value &&
+                !is_reference_wrapper<typename std::decay<Ptr>::type>::value &&
+                !std::is_base_of<Base, typename std::decay<Ptr>::type>::value,
+                decltype((*std::forward<Ptr>(ptr)).*pmd)>::type
+        {
+            return (*std::forward<Ptr>(ptr)).*pmd;
+        }
+
+        template<
+            typename F,
+            typename... Args
+        >
+        constexpr auto INVOKE(
+            F && f,
+            Args&&... args
+        )
+        noexcept(noexcept(std::forward<F>( f )( std::forward<Args>( args )...)))
         -> typename std::enable_if<
                 !std::is_member_pointer<typename std::decay<F>::type>::value,
-                decltype(std::forward<F>(fn)(std::forward<Args>(args)...))>::type
+                decltype(std::forward<F>(f)(std::forward<Args>(args)...))>::type
         {
-            return std::forward<F>(fn)(std::forward<Args>(args)...);
+            return std::forward<F>( f )( std::forward<Args>( args )...);
         }
     }
 
     template< typename F, typename... Args>
-    auto invoke(F && f, Args &&... args)
+    constexpr auto invoke(F && f, Args &&... args)
     /*!
         ensures
             - identical to std::invoke(std::forward<F>(f), std::forward<Args>(args)...)
             - works with C++11 onwards
     !*/
+    noexcept(noexcept(detail::INVOKE(std::forward<F>( f ), std::forward<Args>( args )...)))
```

Comment:
```suggestion
    noexcept(detail::INVOKE(std::forward<F>( f ), std::forward<Args>( args )...))
```

Do there need to be two `noexcept` keywords here?

Reply:
I thought you needed 2: 1 which is the noexcept specifier and the other which is the noexcept operator. The operator is used to determine at compile-time whether the expression is noexcept or not, and the specifier marks the function conditionally noexcept based on a true or false statement.

Reply:
Oh right, duh  :)